functions/CheckForCustomer.js

```diff
@@ -1,148 +1,217 @@
-    invalid = true;
-    invalidMsg = "channelProfile.customerBusinessReferences is not an array"
-  } else if (channelProfile.customerBusinessReferences.length === 0) {
-    invalid = true;
-    invalidMsg = "channelProfile.customerBusinessReferences is empty"
-  }
-
-  //If a sales order document was not passed in, the request is invalid
-  if (!payload) {
-    invalid = true;
-    invalidMsg = "payload was not provided"
-  } else if (!payload.doc) {
-    invalid = true;
-    invalidMsg = "payload.doc was not provided";
-  }
-
-  //If callback is not a function
-  if (!callback) {
-    throw new Error("A callback function was not provided");
-  } else if (typeof callback !== 'function') {
-    throw new TypeError("callback is not a function")
-  }
-
-  if (!invalid) {
-    // Using request for example - A different npm module may be needed depending on the API communication is being made to
-    // The `soap` module can be used in place of `request` but the logic and data being sent will be different
-    let request = require('request');
-
-    let url = "https://localhost/";
-
-    // Add any headers for the request
-    let headers = {
+'use strict'
 
-    };
+let CheckForCustomer = function (ncUtil, channelProfile, flowContext, payload, callback) {
+    const _ = require('lodash');
+    const soap = require('strong-soap/src/soap');
+    const jsonata = require('jsonata');
+    const nc = require('../util/common');
 
-    // Log URL
-    log("Using URL [" + url + "]", ncUtil);
+    let soapClient = null;
 
-    // Set options
-    let options = {
-      url: url,
-      method: "GET",
-      headers: headers,
-      body: payload.doc,
-      json: true
+    let out = {
+        ncStatusCode: null,
+        payload: {}
     };
 
-    try {
-      // Pass in our URL and headers
-      request(options, function (error, response, body) {
-        if (!error) {
-          // If no errors, process results here
-          if (response.statusCode == 200) {
-            if (body.customers && body.customers.length == 1) {
+    if (!callback) {
+        throw new Error("A callback function was not provided");
+    } else if (typeof callback !== 'function') {
+        throw new TypeError("callback is not a function")
+    }
+
+    validateFunction()
+        .then(createSoapClient)
+        .then(searchForCustomer)
+        .then(buildResponse)
+        .catch(handleError)
+        .then(() => callback(out))
+        .catch(error => {
+            logError(`The callback function threw an exception: ${error}`);
+            setTimeout(() => {
+                throw error;
+            });
+        });
+
+    function logInfo(msg) {
+        nc.log(msg, "info");
+    }
+
+    function logWarn(msg) {
+        nc.log(msg, "warn");
+    }
+
+    function logError(msg) {
+        nc.log(msg, "error");
+    }
+
+    async function validateFunction() {
+        let messages = [];
+
+        if (!ncUtil)
+            messages.push("ncUtil was not provided");
```

Comment:
This may be a style/preference thing, but should these if/else if blocks have { brackets around them?  Or are they not required for one liners (like c#)?